src/main/scala/TypeCheck.scala

```diff
@@ -0,0 +1,138 @@
+              env2.updateBind(id -> env(id).consumeBank(i, v % dims(i)._2))
+            case (t, _) => throw InvalidIndex(id, t)
+          }
+        })
+      }
+      case t => throw UnexpectedType(expr.pos, "array access", s"$t[]", t)
+    }
+  }
+
+  private def checkC(cmd: Command)(implicit env: Env, its: ItStack): Env = cmd match {
+    case CPar(c1, c2) => checkC(c2)(checkC(c1), its)
+    case CIf(cond, cons) => {
+      val (cTyp, e1) = checkE(cond)(env.addScope, its)
+      if (cTyp != TBool()) {
+        throw UnexpectedType(cond.pos, "if condition", TBool().toString, cTyp)
+      } else {
+        checkC(cons)(e1, its).endScope._1
+      }
+    }
+    case CUpdate(lhs, rhs) => {
+      val (t1, e1) = checkE(lhs)
+      val (t2, e2) = checkE(rhs)(e1, its)
+      // :< is a subtyping method we defined on the type trait.
+      if (t2 :< t1) (t1, t2, lhs) match {
+        // Reassignment of static ints upcasts to bit<32>
+        case (TStaticInt(_), TStaticInt(_), EVar(id)) =>
+          e2.updateBind(id -> Info(id, TSizedInt(32)))
+        case (TStaticInt(_), TSizedInt(_), EVar(id)) =>
+          e2.updateBind(id -> Info(id, t2))
+        case _ => e2
+      }
+      else throw UnexpectedSubtype(rhs.pos, "assignment", t1, t2)
+    }
+    case CReduce(rop, l, r) => {
+      val (t1, e1) = checkE(l)
+      val (ta, e2) = checkE(r)(e1, its)
+      (t1, ta) match {
+        case (t1, TArray(t2, dims)) =>
+          if (t2 :< t1 == false || dims.length != 1 || dims(0)._1 != dims(0)._2) {
+            throw ReductionInvalidRHS(r.pos, rop, t1, ta)
+          } else {
+            e2
+          }
+        case _ =>
+          throw ReductionInvalidRHS(r.pos, rop, t1, ta)
+      }
+    }
+    case l@CLet(id, typ, exp) => {
+      val (t, e1) = checkE(exp)
+      typ match {
+        case Some(t2) if t :< t2 => e1.addBind(id -> Info(id, t2))
+        case Some(t2) => throw UnexpectedType(exp.pos, "let", t.toString, t2)
+        case None => l.typ = Some(t); e1.addBind(id -> Info(id, t))
+      }
+    }
+    case CFor(range, par, combine) => {
+      val iter = range.iter
+      // Add binding for iterator in a separate scope.
+      val e1 = env.addScope.addBind(iter -> Info(iter, range.idxType)).addScope
+      // Check for body and pop the scope.
+      val (e2, binds) = checkC(par)(e1, iter :: its).endScope
+      // Create scope where ids bound in the parallel scope map to fully banked arrays.
+      val vecBinds = binds.map({ case (id, inf) =>
+        id -> Info(id, TArray(inf.typ, List((range.u, range.u))))
+      })
+      // Remove the binding of the iterator and add the updated vector bindings.
+      checkC(combine)(e2.endScope._1.addScope ++ vecBinds, its).endScope._1
+    }
+    case CExpr(e) => checkE(e)._2
+    case CEmpty => env
+    case CSeq(c1, c2) => {
+      val _ = checkC(c1)
+      val e2 = checkC(c2)(env, its)
+      // FIXME(rachit): This should probably be intersection of e1 and e2
+      e2
```

Comment:
Hmm, I'm not sure the intersection is quite what we want: if `c1` and `c2` both declare new variables, for example, both of those variables should be available (together) in the context that follows. (But maybe you're thinking of something else here.)

Reply:
I see your point. This is more subtle. Opened #51 to discuss this.